### DISCUSSION TITLE
How to insure a provider with autoDispose + family is disposed before another is initialized

### DISCUSSION THREAD

**Discussion by trevhud** [rrousselGit/riverpod#909] [Q&A]
**Describe what scenario you think is uncovered by the existing examples/articles**

I have a autoDispose family future provider that is watched in a widget. When the family argument changes, the widget is rebuild with the new argument and the new instance of the provider is initialized. After that, the previous instance referenced by the previous argument is disposed. It is unclear how I would make sure that onDispose is called on the first instance before the second is created.

**Describe why existing examples/articles do not cover this case**

The family and autoDispose documentation does not describe this potential problem.

**Comment by rrousselGit** [comment]
What you're asking for most likely stems from a misunderstanding of what `family` does.   

Could you explain clearly why you want this?

**Reply by trevhud** [reply]
@rrousselGit My particular use case is around using the camera with google's on device machine learning library. I use a family to initialize and provide the camera at a low resolution for ocr, and at a high resolution for barcode scanning. Thus I need to be able to pass in a parameter to retrieve the correct one.

But when the widget tree changes and the watch for one is replaced with the other, the new one tries to initialize before the other is disposed.

I would think that when a watch is and replaced with another, that provider's onDispose would get called before the new provider is initialized. I would also think that onDispose should be able to accept a FutureOr such that you could insure that all your asynchronous dispose functions get run before that same provider sibling gets created.

Thanks for your help, definitely open to suggestions on other ways to accomplish this.

**Reply by rrousselGit** [reply]
That's not quite answering my question: Why do you need the "new" provider to be disposed of before the "previous" one?

`family` isn't working like you're suggesting it is. It's not a "previous/new" instance – it is entirely possible that both instances will be active simultaneously.  
As such, asking for an instance to be created only after another one was disposed doesn't make sense, since they are completely independent.

**Reply by trevhud** [reply]
@rrousselGit 
It's that I need the previous provider to be disposed of before the current one is created. The reason is because you can only use one camera controller at a time and it must be created with the correct resolution, based on a parameter.

I understand that both can be active simultaneously and that they are independent. This makes sense and works for many use cases. But what would you suggest for when you don't want this to be the case and yet you need to pass a parameter into the provider?

**Reply by rrousselGit** [reply]
>  The reason is because you can only use one camera controller at a time and it must be created with the correct resolution, based on a parameter.

In this case, using `family` does not make sense, since `family` is about possibly having multiple controllers at once.

Rather, you may want something like:

```dart
final dimensionProvider = StateProvider((ref) => Size(600, 400));

final cameraControllerProvider = Provider((ref) {
  final dimension = ref.watch(dimensionProvider);

  return CameraController(dimension);
});
```

and when you want to change the resolution, instead do `ref.read(dimensionProvider.notifier).state = Size(...)`;

**Reply by trevhud** [reply]
Yeah, that's how I had it initially. I was hoping to pass in the parameter so that this camera system could be fully encapsulated. That way I can use it elsewhere with completely different params. I also have to pass in the callback that needs to get called with the result of the image detection and processing. So I was hoping to not need to manage that state provider. Oh well, let me know if you happen to think of any other ideas.

**Reply by TimWhiting** [reply]
If you make the cameracontroller a ChangeNotifier or StateNotifier you can create a setter / method to change the dimensions. Also why are you using callbacks rather than Futures? The Notifier could have a method to return a future or a stream with the result of the image detection / processing.

**Reply by trevhud** [reply]
@rrousselGit I was able to refactor to use a non-family provider. However, I still have the same problem that my provider gets recreated before the previous onDispose is done running. Seems like it stems from onDispose not being able to take a future. What do you think?
